deepchem/feat/protein_tokenizer.py

```diff
@@ -0,0 +1,155 @@
+  that are terminated on both ends with [CLS] and [SEP] annotations, and
+  converts each string into a matching array of integer tokens.
+
+  The tokenizer is case-insensitive and safely ignores all non-FASTA characters
+  (spaces, newlines, commas, etc.).
+
+  eg. ["EMV[CLS]ABCDE[SEP]X", "WMY[CLS]FGH[SEP]XAB"] -> [[0, 1, 2, 3, 4], [5, 6, 7]]
+  """
+
+  def featurize(self,
+                proteins: Iterable[str],
+                log_every_n: int = 1000) -> np.ndarray:
+    """Tokenizes protein sequences.
+
+    Parameters
+    ----------
+    proteins: Iterable[str]
+      An iterable (list, array, etc.) of FASTA-format protein sequence strings
+      eg. ["ABCD[CLS]EFGHIJK[SEP]LMNOP", "[CLS]A[SEP]"]
+      Newlines, spaces, and commas in protein sequence strings are safely ignored.
+    log_every_n: int, optional (default 1000)
+      How many proteins are tokenized every time a tokenization is logged.
+
+    Returns
+    -------
+    np.ndarray
+      An array of arrays of integer tokens (one array for every sequence).
+    """
+    # Calls featurize() in parent class, which will call _featurize() for each protein in proteins
+    return Featurizer.featurize(self, proteins, log_every_n)
+
+  def untransform(self, input_sequences: Iterable[Iterable[int]]) -> tuple:
+    """Convert from tokenized arrays back into original string.
+
+    Parameters
+    ----------
+    input_sequences: Iterable[int]
+      Iterable of the token arrays to be untransformed.
+
+    Returns
+    -------
+    tuple
+      Tuple of strings, where each element is a protein sequence.
+    """
+    acid_codes = list(FASTA_tokens.keys())  # List of all keys in FASTA_tokens
+    all_tokens = list(
+        FASTA_tokens.values())  # List of all values in FASTA_tokens
+    output_acid_codes: tuple = tuple(
+    )  # FASTA amino acid codes for values in input
+    # Iterating through input_tokens
+    for sequence in input_sequences:
+      token_codes = "[CLS]"
+      for token in sequence:
+        idx = all_tokens.index(token)  # Get index of token
+        code = acid_codes[idx]  # Get corresponding key (FASTA amino acid code)
+        token_codes = token_codes + code
+      token_codes = token_codes + "[SEP]"
+      output_acid_codes = output_acid_codes + (token_codes,)
+    return output_acid_codes
+
+  def _extract_relevant_sequence(self, protein: str) -> str:
+    """Extracts the relevant part of a protein sequence between the first
+    [CLS] mark and the first [SEP] mark following the first [CLS] mark.
+
+    eg. "AB[CLS]ABCD[SEP]A" -> "ABCD"
+
+    Parameters
+    ----------
+    protein: str
+      A FASTA-format protein sequence with start and end annotations.
+
+    Returns
+    -------
+    str
+      A FASTA-format protein sequence without start and end annotations.
+    """
+    first_idx = -1  # Will be set to first string index after [CLS]
+    for i in range(len(protein) - 4):
+      if (first_idx == -1 and protein[i:i + 5] == "[CLS]"):
```

Comment:
I think this could break in edge cases. For example consider a string where there is no [CLS]. We'll get a string out of bounds exception

Reply:
To fix this, you want an explicit check you're not violating string bounds

Reply:
Or maybe I could just initialize first_idx to 0 instead of -1?

If the loop is entered, index 0 cannot be out of bounds. Further, there's no way that the correct value of first_idx is actually 0 in a validly annotated FASTA string since [CLS] is already five characters.

Reply:
I just made that change. I think that should make an explicit check redundant? :)